src/extension/infer.rs

```diff
@@ -663,31 +680,85 @@ impl UnificationContext {
         self.results()
     }
 
-    /// Instantiate all variables in the graph with the empty extension set.
+    /// Gather all the transitive dependencies (induced by constraints) of the
+    /// variables in the context.
+    fn search_variable_deps(&self) -> HashSet<Meta> {
+        let mut variable_scope: HashSet<Meta> = self.variables.clone();
+        let mut scope_size = 0;
+        let mut new_scope_size = variable_scope.len();
+        while new_scope_size > scope_size {
+            scope_size = new_scope_size;
+            for m in variable_scope.clone().iter() {
+                for c in self.get_constraints(m).unwrap().iter() {
```

Comment:
So it's not clear to me from this what happens when we call `self.get_constraints(m)` for an `m` *that we've seen before* (on a previous iteration of the outermost loop). Will we get anything back? If so, is there any chance that we'll actually get a new thing to add to `variable_scope`? Or is it only the new/never-seen-before values of `m` that lead to finding new things? The latter would allow us to rewrite this loop much more simply but I don't know whether it's true....

Reply:
Ok, so if each iteration made a set of new stuff to process in the next iteration, you could do away with the sizes and just have two sets - something like
```
let mut seen_variable_scope = HashSet::new();
let mut new_variables = self.variables.clone();
while !new_variables.empty() {
   let mut newer_variables = HashSet::new();
   for m in new_variables.iter() {
      ....add to newer_variables....
   }
   seen_variable_scope.extend(new_variables);
   new_variables = newer_variables
}
```
Moreover, this'll change your algorithm from O(n^2) to something closer to linear, so....

Reply:
I've gone for the former